File: src/types/edit.ts
/**
 * Types pour le module d'édition non-destructive (Phase 4.1 — Event Sourcing Engine)
 */

// Liste exhaustive des types d'événements d'édition supportés
export const EDIT_EVENT_TYPES = [
  'EXPOSURE',
  'CONTRAST',
  'SATURATION',
  'HIGHLIGHTS',
  'SHADOWS',
  'WHITES',
  'BLACKS',
  'CLARITY',
  'VIBRANCE',
  'TEMPERATURE',
  'TINT',
  'SHARPNESS',
  'NOISE_REDUCTION',
  'CROP',
  'ROTATION',
] as const;

export type EditEventType = (typeof EDIT_EVENT_TYPES)[number];

/** Payload JSON d'un événement d'édition */
export interface EditPayload {
  param: string;
  value: number;
  prev_value?: number;
}

/** DTO d'un événement d'édition (retourné par get_edit_history) */
export interface EditEventDTO {
  id: number;
  event_type: EditEventType;
  payload: EditPayload;
  is_undone: boolean;
  created_at: string;
}

/**
 * DTO de l'état courant d'édition d'une image.
 * `state` est un dictionnaire parametre/valeur reconstruit par rejeu des events.
 */
export interface EditStateDTO {
  image_id: number;
  state: Record<string, number>;
  can_undo: boolean;
  can_redo: boolean;
  event_count: number;
}

/** DTO d'un snapshot nommé (Phase 4.3) */
export interface SnapshotDTO {
  id: number;
  image_id: number;
  name: string;
  description?: string;
  event_count: number;
  created_at: string;
}

File: src-tauri/src/models/edit.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Raw edit event as stored in DB (used in service layer for DB queries)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[allow(dead_code)]
pub struct EditEvent {
    pub id: i64,
    pub image_id: i64,
    pub event_type: String,
    pub payload: serde_json::Value, // JSON: { "param": "exposure", "value": 0.75 }
    pub is_undone: bool,
    pub session_id: Option<String>,
    pub created_at: String,
}

/// Snapshot of computed edit state for a given image (performance optimization)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[allow(dead_code)]
pub struct EditSnapshot {
    pub image_id: i64,
    pub snapshot: serde_json::Value, // Map<param, value>
    pub event_count: i64,
    pub updated_at: String,
}

/// DTO exposé au frontend : état courant de l'édition d'une image
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EditStateDTO {
    pub image_id: i64,
    pub state: HashMap<String, f64>, // { "exposure": 0.5, "contrast": -0.2, ... }
    pub can_undo: bool,
    pub can_redo: bool,
    pub event_count: i64,
}

/// DTO exposé au frontend : un événement d'édition
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EditEventDTO {
    pub id: i64,
    pub event_type: String,
    pub payload: serde_json::Value,
    pub is_undone: bool,
    pub created_at: String,
}
